### PR TITLE
SOL-53 [Integrate] Integrate smart contract function for viewing specific campaigns

### PR DESCRIPTION
**Pull request by wylenjoan**
## Issue Link
- https://framgiaph.backlog.com/view/SPH_SOLIDITY-53

## Target Page | Component | Function
- `/campaigns/<campaign-id>`

## Notes
ℹ️ This PR contains changes in smart contract to include the donations in getting specific campaign.

## Pre-condition
- N/A

### For BE / Smart Contract

1. Update smart_contract/.env
```
API_URL = "https://eth-sepolia.g.alchemy.com/v2/KPjBlFd432ifoUNi8UKzGcq-uL1OhX31"
USER_KEY = <your metamask private key>
CONTRACT_ADDRESS = "0xbDCd2efc543c5996F17cDA26d6F7197D730F81EE"
```
2. In root directory, run `npm run hardhat:compile`
3. Create campaign
   - Using the Menu: `npm run hardhat:interact-local` in root directory
      image
      - Choose [option 1]
      - Note: 
   - Using the script command in `smart_contract` directory:
   ```
   npx ts-node scripts/createCampaign.ts <name> <title> <story> <goal> <deadline>
   
   # Example
   npx ts-node scripts/createCampaign.ts "sample name" "sample title" "sample story" 36510000000 1701360000
   ```
4. Get the campaign with donations (NOTE: it won't show donations if there are no campaign donations)
   - Using the Menu: 
     - Choose [option 3]
   - Using the script command in `smart_contract` directory:
   ```
   npx ts-node scripts/getCampaign.ts <campaign id>

   # Example
   npx ts-node scripts/getCampaign.ts 0
   ```

### For FE
1. Go to `/campaigns/<campaign-id>`
   - You can go to `/profile` and click on one of the campaign cards to redirect to `/campaigns/<campaign-id>` page

## Expected Output
- The details of the campaign should be fetched and shown in the FE
- The edit button near Campaign Title should show only when the logged in user is the creator.

## Definition of Done
- The campaign details is fetched and shown dynamically in `/campaigns/<campaign-id>` page

## Test View Points
- [ ] Add here

## Screenshots | Recordings
<details>
<summary> Get specific campaign with donations in smart contract </summary>

image

</details>

<details>
<summary> View specific campaign in FE </summary>

image

</details>

### General
- [ ] Does the code meet the requirements of the project?
- [ ] Is the code readable and maintainable?
- [ ] Are there any unnecessary or redundant lines of code?
- [ ] Are there any long methods or functions that should be refactored?
- [ ] Are there any performance issues?

### Functionality
- [ ] Are there any logical errors?
- [ ] Are edge cases handled properly?
- [ ] Are there any potential security vulnerabilities?

### Code Style
- [ ] Does the code adhere to the project's coding standards?
- [ ] Are variable and function names descriptive and meaningful?
- [ ] Is there proper use of comments and documentation?
- [ ] Is the code properly formatted? Are indentation, spacing, and formatting consistent?

### Security
- [ ] Is the website secure from common attacks?
- [ ] Are inputs properly validated and sanitized?
- [ ] Are sensitive data and credentials properly protected?

### Testing
- [ ] Are there sufficient test cases?
- [ ] Do the test cases cover all functionality?
- [ ] Are the test cases properly written and documented?

### Additional Comments
- [ ] Are there any other issues or concerns with the code?
- [ ] Are there any suggestions for improvement?
- [ ] Are there any positive aspects of the code that should be highlighted?

### CODE REVIEW

**Review by kenorenzsun — DISMISSED** [review]

**Review by johnpaul-sun — DISMISSED** [review]
Approved ✅

Review Time: 10 mins
Comments: 0
IT: 0

File: fe/src/pages/campaigns/[id].vue [hunk 1]
```diff
@@ -40,20 +47,81 @@ import { storeToRefs } from "pinia";
...
+              "https://img.freepik.com/free-psd/3d-illustration-person_23-2149436182.jpg",
+          },
+          imgSrc:
+            "https://images.unsplash.com/photo-1529390079861-591de354faf5",
+          story: result[0][4],
+          daysLeft: getDaysLeft(result[0][7]),
+          totalDonation: Number(ethers.formatEther(result[0][6])),
+          totalBackers: Number(result[0][8]),
+          donations: [],
+        };
```
Comment (chevyB) [inline fe/src/pages/campaigns/[id].vue]: is there a way we can construct this in BE?

File: fe/src/pages/campaigns/[id].vue [hunk 2]
```diff
@@ -40,20 +47,81 @@ import { storeToRefs } from "pinia";
...
+  donations: [],
+});
+
+const getCampaign = async (id: number): Promise<void> => {
+  isLoading.value = true;
+  const smartContract = await getSmartContract();
+  if (smartContract !== null) {
+    smartContract
+      .getCampaign(id)
+      .then((result) => {
```
Comment (chevyB) [inline fe/src/pages/campaigns/[id].vue]: seems there is a lot of logics, and flow is explicit.
Lets try to use implicit approach

- use try catch
- Creating a function for each logic that can describe what do you want to structure

```
const getCampaign = (data) => {
 return  {
          campaignId: Number(data[0][0]),
          title: data[0][3],
          creator: {
            address: data[0][1],
            fullName: data[0][2],
            imgSrc:
              "https://img.freepik.com/free-psd/3d-illustration-person_23-2149436182.jpg",
          },
          imgSrc:
            "https://images.unsplash.com/photo-1529390079861-591de354faf5",
          story: data[0][4],
          daysLeft: getDaysLeft(data[0][7]),
          totalDonation: Number(ethers.formatEther(data[0][6])),
          totalBackers: Number(data[0][8]),
          donations: [],
        };
}

try {
  const result = smartContract.getCampaign(id)
  const campaign = getCampaign(result)
} catch {

}
```

File: fe/src/pages/campaigns/[id].vue [hunk 3]
```diff
@@ -40,20 +47,81 @@ import { storeToRefs } from "pinia";
...
+            address: donation[2].toString(),
+            amount: Number(ethers.formatEther(donation[3])),
+          })
+        );
+        if (
+          address.value.toLowerCase() ===
+          campaign.value.creator.address.toLowerCase()
+        ) {
+          canEdit.value = true;
+        }
```
Comment (chevyB) [inline fe/src/pages/campaigns/[id].vue]: seems we are vulnerable if we dont process this in BE

**Comment by ipilikent** [comment]
## Approved PR
- The user can view specific campaigns
- The details of the campaign can fetch the user and shown in the FE
- The edit button near Campaign Title is displayed for author of the campaign
- The edit button not displayed when not author of the campaign 

## Proof
<details>

- Get specific campaign with donations in smart contract
image

- View specific campaign in FE
image

</details>

Test Case: https://docs.google.com/spreadsheets/d/114GrEn-LQQ8saIvjz9zqenuk05qmO4dN/edit#gid=1122662191
-----
- Total Test time: 0.75 hr.
- Total Comments: 0
- Total Iterations: 0

**Review by ipilikent — APPROVED** [review]

File: fe/src/pages/campaigns/[id].vue (hunk 1 above)
Comment (wylenjoan) [reply fe/src/pages/campaigns/[id].vue]: @chevyB the data from BE looks like this:
- `result[0]` is for the Campaign details
- `result[1]` is the array of Donations of the Campaign

image

File: fe/src/pages/campaigns/[id].vue (hunk 3 above)
Comment (wylenjoan) [reply fe/src/pages/campaigns/[id].vue]: @chevyB this condition hides/shows the edit icon beside the title. it is based on comparing the address of the currently logged in user and the campaign creator's address 🙇 

image

File: fe/src/pages/campaigns/[id].vue (hunk 2 above)
Comment (wylenjoan) [reply fe/src/pages/campaigns/[id].vue]: See new commit: [Restructure code](https://github.com/framgia/sph-solidity/pull/36/commits/a633718a2bb99fd10727f521e9307acba65b8bf7)

**Review by chevyB — DISMISSED** [review]

**Review by kenorenzsun — APPROVED** [review]

**Review by arden-dave-sun — APPROVED** [review]

**Pull request merged**